lib/pedicel/validator.rb

```diff
@@ -1,70 +1,106 @@
+  # https://developer.apple.com/library/content/documentation/PassKit/Reference/PaymentTokenJSON/PaymentTokenJSON.html
+  # This purposefully only does syntactic validation (as opposed to semantic).
   class Validator
     class Error < StandardError; end
     class TokenFormatError < Error; end
     class TokenDataFormatError < Error; end
 
-    DRY_CUSTOM_PREDICATE_ERRORS = {
-      base64?: 'invalid base64',
-      hex?: 'invalid hex',
-      pan?: 'invalid pan',
-      yymmdd?: 'invalid date format YYMMDD',
-    }
-
     module Predicates
       include Dry::Logic::Predicates
 
-      predicate(:base64?) { |value| !Base64.decode64(value).nil? rescue false }
+      CUSTOM_PREDICATE_ERRORS = {
+        base64?:          'invalid base64',
+        hex?:             'invalid hex',
+        pan?:             'invalid pan',
+        yymmdd?:          'invalid date format YYMMDD',
+        ec_public_key?:   'is not an EC public key',
+        pkcs7_signature?: 'is not a PKCS7 Signature',
+        eci?:             'not an ECI indicator',
+        hex_sha256?:      'not hex-encoded SHA256',
+        base64_sha256?:   'not base64-encoded SHA256',
+      }.freeze
+
+      match_b = String.new.respond_to?(:match?) ? lambda{|s, re| s.match?(re)} : lambda{|s, re| !!(s =~ re)}
+      # Support Ruby 2.3, but use the faster #match? when available.
+
+      predicate(:base64?) do |x|
+        str?(x) &&
+          match_b.(x, /\A[=A-Za-z0-9+\/]*\z/) && # allowable chars
+          x.length.remainder(4).zero? && # multiple of 4
+          !match_b.(x, /=[^$=]/) && # may only end with ='s
+          !match_b.(x, /===/) # at most 2 ='s
+      end
+
+      #predicate(:strict_base64?) { |x| !!Base64.strict_decode64(x) rescue false }
 
-      predicate(:hex?) { |value| !Regexp.new(/\A[a-f0-9-]*\z/i).match(value).nil? }
+      predicate(:base64_sha256?) { |x| base64?(x) && Base64.decode64(x).length == 32 }
 
-      predicate(:pan?) { |value| !Regexp.new(/\A[0-9]{13,19}\z/).match(value).nil? }
+      predicate(:hex?) { |x| str?(x) && match_b.(x, /\A[a-f0-9]*\z/i) }
 
-      predicate(:yymmdd?) do |value|
-        return false unless value.length == 6
-        Time.new(value[0..1],value[2..3], value[4..5]).is_a?(Time) rescue false
-      end
+      predicate(:hex_sha256?) { |x| hex?(x) && x.length == 64 }
+
+      predicate(:pan?) { |x| str?(x) && match_b.(x, /\A[1-9][0-9]{11,18}\z/) }
+
+      predicate(:yymmdd?) { |x| str?(x) && match_b.(x, /\A\d{6}\z/) }
+
+      predicate(:eci?) { |x| str?(x) && match_b.(x, /\A\d{2}\z/) }
+
+      predicate(:ec_public_key?) { |x| base64?(x) && OpenSSL::PKey::EC.new(Base64.decode64(x)).check_key rescue false }
+
+      predicate(:pkcs7_signature?) { |x| base64?(x) && !!OpenSSL::PKCS7.new(Base64.decode64(x)) rescue false }
     end
 
     TokenSchema = Dry::Validation.Schema do
       configure do
+        # NOTE: This option removes/sanitizes hash element not mentioned/tested.
+        # Hurray for good documentation.
```

Comment:
I think it's hooray or hurrah :smile:

Reply:
My dict doesn't complain. Also https://en.wikipedia.org/wiki/Hurray, but I have no feeling about this. However, I think the comment should be improved. I don't really understand what is achieved by what. Perhaps @kse-clearhaus can clarify (and push a fix if necessary)? 🙂

Reply:
Hmm, you are right. I think I googled it without any results... Weird. I will crawl back into my cave.